pam-1.1.1/modules/pam_issue/pam_issue.8
.\"     Title: pam_issue
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.74.0 <http://docbook.sf.net/>
.\"      Date: 06/16/2009
.\"    Manual: Linux-PAM Manual
.\"    Source: Linux-PAM Manual
.\"  Language: English
.\"
.TH "PAM_ISSUE" "8" "06/16/2009" "Linux-PAM Manual" "Linux\-PAM Manual"
.\" -----------------------------------------------------------------
.\" * (re)Define some macros
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" toupper - uppercase a string (locale-aware)
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de toupper
.tr aAbBcCdDeEfFgGhHiIjJkKlLmMnNoOpPqQrRsStTuUvVwWxXyYzZ
\\$*
.tr aabbccddeeffgghhiijjkkllmmnnooppqqrrssttuuvvwwxxyyzz
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" SH-xref - format a cross-reference to an SH section
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de SH-xref
.ie n \{\
.\}
.toupper \\$*
.el \{\
\\$*
.\}
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" SH - level-one heading that works better for non-TTY output
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de1 SH
.\" put an extra blank line of space above the head in non-TTY output
.if t \{\
.sp 1
.\}
.sp \\n[PD]u
.nr an-level 1
.set-an-margin
.nr an-prevailing-indent \\n[IN]
.fi
.in \\n[an-margin]u
.ti 0
.HTML-TAG ".NH \\n[an-level]"
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
\." make the size of the head bigger
.ps +3
.ft B
.ne (2v + 1u)
.ie n \{\
.\" if n (TTY output), use uppercase
.toupper \\$*
.\}
.el \{\
.nr an-break-flag 0
.\" if not n (not TTY), use normal case (not uppercase)
\\$1
.in \\n[an-margin]u
.ti 0
.\" if not n (not TTY), put a border/line under subheading
.sp -.6
\l'\n(.lu'
.\}
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" SS - level-two heading that works better for non-TTY output
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de1 SS
.sp \\n[PD]u
.nr an-level 1
.set-an-margin
.nr an-prevailing-indent \\n[IN]
.fi
.in \\n[IN]u
.ti \\n[SN]u
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.ps \\n[PS-SS]u
\." make the size of the head bigger
.ps +2
.ft B
.ne (2v + 1u)
.if \\n[.$] \&\\$*
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" BB/BE - put background/screen (filled box) around block of text
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de BB
.if t \{\
.sp -.5
.br
.in +2n
.ll -2n
.gcolor red
.di BX
.\}
..
.de EB
.if t \{\
.if "\\$2"adjust-for-leading-newline" \{\
.sp -1
.\}
.br
.di
.in
.ll
.gcolor
.nr BW \\n(.lu-\\n(.i
.nr BH \\n(dn+.5v
.ne \\n(BHu+.5v
.ie "\\$2"adjust-for-leading-newline" \{\
\M[\\$1]\h'1n'\v'+.5v'\D'P \\n(BWu 0 0 \\n(BHu -\\n(BWu 0 0 -\\n(BHu'\M[]
.\}
.el \{\
\M[\\$1]\h'1n'\v'-.5v'\D'P \\n(BWu 0 0 \\n(BHu -\\n(BWu 0 0 -\\n(BHu'\M[]
.\}
.in 0
.sp -.5v
.nf
.BX
.in
.sp .5v
.fi
.\}
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" BM/EM - put colored marker in margin next to block of text
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de BM
.if t \{\
.br
.ll -2n
.gcolor red
.di BX
.\}
..
.de EM
.if t \{\
.br
.di
.ll
.gcolor
.nr BH \\n(dn
.ne \\n(BHu
\M[\\$1]\D'P -.75n 0 0 \\n(BHu -(\\n[.i]u - \\n(INu - .75n) 0 0 -\\n(BHu'\M[]
.in 0
.nf
.BX
.in
.fi
.\}
..
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "Name"
pam_issue \- PAM module to add issue file to user prompt
.SH "Synopsis"
.fam C
.HP \w'\fBpam_issue\&.so\fR\ 'u
\fBpam_issue\&.so\fR [noesc] [issue=\fIissue\-file\-name\fR]
.fam
.SH "DESCRIPTION"
.PP
pam_issue is a PAM module to prepend an issue file to the username prompt\&. It also by default parses escape codes in the issue file similar to some common getty\'s (using \ex format)\&.
.PP
Recognized escapes:
.PP
\fB\ed\fR
.RS 4
current day
.RE
.PP
\fB\el\fR
.RS 4
name of this tty
.RE
.PP
\fB\em\fR
.RS 4
machine architecture (uname \-m)
.RE
.PP
\fB\en\fR
.RS 4
machine\'s network node hostname (uname \-n)
.RE
.PP
\fB\eo\fR
.RS 4
domain name of this system
.RE
.PP
\fB\er\fR
.RS 4
release number of operating system (uname \-r)
.RE
.PP
\fB\et\fR
.RS 4
current time
.RE
.PP
\fB\es\fR
.RS 4
operating system name (uname \-s)
.RE
.PP
\fB\eu\fR
.RS 4
number of users currently logged in
.RE
.PP
\fB\eU\fR
.RS 4
same as \eu except it is suffixed with "user" or "users" (eg\&. "1 user" or "10 users")
.RE
.PP
\fB\ev\fR
.RS 4
operating system version and build date (uname \-v)
.RE
.SH "OPTIONS"
.PP
.PP
\fBnoesc\fR
.RS 4
Turns off escape code parsing\&.
.RE
.PP
\fBissue=\fR\fB\fIissue\-file\-name\fR\fR
.RS 4
The file to output if not using the default\&.
.RE
.SH "MODULE TYPES PROVIDED"
.PP
Only the
\fBauth\fR
module type is provided\&.
.SH "RETURN VALUES"
.PP
.PP
PAM_BUF_ERR
.RS 4
Memory buffer error\&.
.RE
.PP
PAM_IGNORE
.RS 4
The prompt was already changed\&.
.RE
.PP
PAM_SERVICE_ERR
.RS 4
A service module error occurred\&.
.RE
.PP
PAM_SUCCESS
.RS 4
The new prompt was set successfully\&.
.RE
.SH "EXAMPLES"
.PP
Add the following line to
\FC/etc/pam\&.d/login\F[]
to set the user specific issue at login:
.sp
.if n \{\
.RS 4
.\}
.fam C
.ps -1
.nf
.if t \{\
.sp -1
.\}
.BB lightgray adjust-for-leading-newline
.sp -1

        auth optional pam_issue\&.so issue=/etc/issue
      
.EB lightgray adjust-for-leading-newline
.if t \{\
.sp 1
.\}
.fi
.fam
.ps +1
.if n \{\
.RE
.\}
.sp
.SH "SEE ALSO"
.PP

\fBpam.conf\fR(5),
\fBpam.d\fR(5),
\fBpam\fR(8)
.SH "AUTHOR"
.PP
pam_issue was written by Ben Collins <bcollins@debian\&.org>\&.


noflushd/NEWS
* noflushd 2.7.5 (09. May 2005)

  - Fix handling of raid and similar meta devices on 2.6 kernels.

* noflushd 2.7.4 (08. Aug 2004)

  - Fix lousy spindown times on 2.6 kernels.
  - Fix SCSI support on 2.6 kernels.
  - Fix pid detection in generic startup script.
  - Document quirks with hotplugged drives.

* noflushd 2.7.3 (09. Apr 2004)

  - Fix race condition that might leave kupdated/pdflushd disabled
    aftern noflushd has terminated.
  - Fix spindown when all disks are auto-detected.

* noflushd 2.7.2 (16. Mar 2004)

  - A change in 2.7.1 erroneously flushed the buffer cache before
    spindown, causing frequent spinups. This change has been reverted.
    
* noflushd 2.7.1 (07. Mar 2004)

  - Fix immediate spinup on 2.6 kernels.
  - Elevate SCSI support from compile-time to command line option.
  - Bug fixes in the disk statistics parser for 2.6 kernels.
  - Ship kernel patch to track down which processes cause spinups.

* noflushd 2.7 (20. Jan 2004)

  - Support Linux kernel 2.6.x.
  - Default to monitorring all disks unless specific disks are given
    on the command line. This change should yield a sane default
    behaviour for a majority of users without requiring explicit
    configuration.

* noflushd 2.6.3 (13. May 2002)

  - Bugfix for arm and other archs with weird /proc/partitions layout.

* noflushd 2.6.2 (20. Apr 2002)

  - Add IDE raid support.
  - Add timeout logging script.

* noflushd 2.6.1 (03. Feb. 2002)

  - Fix BUG when using noflushd with disks on second IDE channel on 2.4
    kernels.

* noflushd 2.6 (22. Jan 2002)

  - Document noflushd's suckage with any journaling fs, including ext3.
  - New feature prevents spindown if machine is in interactive use.
  - Add hack fix some '0 spindown time' uglies.

* noflushd 2.5 (06. Aug 2001)

  - Make sure hotplugged disks get synced properly.
  - Only sync present partitions. (Fixes high CPU usage bug on devfs
    systems.)
  - Generic startup skript works on latest 2.4 kernels.

* noflushd 2.4 (13. Mar 2001)

  - Timeout lists are back to a working state.
  - Support very latest 2.4 kernels.

* noflushd 2.3 (12. Jan 2001)

  - Support latest 2.4 kernels. (Leaves kernels 2.3.99pre8 to 2.4.0-test8
    unsupported.)
  - Fix multi-disk support with long bdflush timeouts.

* noflushd 2.2 (22. Dec 2000)

  - Fixed bug in spindown code. Now actually spins down. Duh!

* noflushd 2.1 (20. Dec 2000)

  - Disks get synced prior to spindown.
  - Improved devfs support.

* noflushd 2.0 (14. Nov 2000)

  - noflushd now comes with its own syncer that takes over the task from
    kupdate when the first read-write mounted disk is spun down. This
    enhances spindown times in a multidisk environment at the expense of
    less smooth and more bursty background flushing when the noflushd 
    syncer is in use.

  - Conflicting entries in 2.2-style /proc/stat (e.g. /dev/sda and /dev/hda)
    won't cause noflushd to give up anymore. Instead, noflushd will treat
    both disks pretty much like one, i.e. they're either both up or spun
    down, no matter if activity only goes to one of them.

  - Compilation environment switched to automake/autoconf.

  - When installing via 'make install', noflushd is no longer added to
    any runlevel. 

  - noflushd moved to SourceForge. The project homepage is located at
    http://noflushd.sourceforge.net, a mailing list is available via
    noflushd-devel@lists.sourceforge.net. Visit the homepage for 
    subscription information.



pam-0.76/Linux-PAM/modules/pam_issue/pam_issue.c
/* pam_issue module - a simple /etc/issue parser to set PAM_USER_PROMPT
 *
 * Copyright 1999 by Ben Collins <bcollins@debian.org>
 *
 * Needs to be called before any other auth modules so we can setup the
 * user prompt before it's first used. Allows one argument option, which
 * is the full path to a file to be used for issue (uses /etc/issue as a
 * default) such as "issue=/etc/issue.telnet".
 *
 * We can also parse escapes within the the issue file (enabled by
 * default, but can be disabled with the "noesc" option). It's the exact
 * same parsing as util-linux's agetty program performs.
 *
 * Released under the GNU LGPL version 2 or later
 */

#define _GNU_SOURCE
#define _BSD_SOURCE

#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <string.h>
#include <unistd.h>
#include <sys/utsname.h>
#include <utmp.h>
#include <malloc.h>
#include <time.h>

#include <security/_pam_macros.h>

#define PAM_SM_AUTH

#include <security/pam_modules.h>

static int _user_prompt_set = 0;

char *do_prompt (FILE *);

/* --- authentication management functions (only) --- */

PAM_EXTERN
int pam_sm_authenticate(pam_handle_t *pamh, int flags, int argc,
                        const char **argv)
{
    int retval = PAM_SUCCESS;
    FILE *fd;
    int parse_esc = 1;
    char *prompt_tmp = NULL;
    const char *cur_prompt = NULL;
    struct stat st;
    char *issue_file = NULL;

   /* If we've already set the prompt, don't set it again */
    if(_user_prompt_set)
	return PAM_IGNORE;
    else
       /* we set this here so if we fail below, we wont get further
	  than this next time around (only one real failure) */
	_user_prompt_set = 1;

    for ( ; argc-- > 0 ; ++argv ) {
	if (!strncmp(*argv,"issue=",6)) {
	    issue_file = (char *) strdup(6+*argv);
	    if (issue_file != NULL) {
		D(("set issue_file to: %s", issue_file));
	    } else {
		D(("failed to strdup issue_file - ignored"));
		return PAM_IGNORE;
	    }
	} else if (!strcmp(*argv,"noesc")) {
	    parse_esc = 0;
	    D(("turning off escape parsing by request"));
	} else
	    D(("unknown option passed: %s", *argv));
    }

    if (issue_file == NULL)
	issue_file = strdup("/etc/issue");

    if ((fd = fopen(issue_file, "r")) != NULL) {
	int tot_size = 0;

	if (fstat(fileno(fd), &st) < 0)
	    return PAM_IGNORE;

	retval = pam_get_item(pamh, PAM_USER_PROMPT,
			      (const void **) &cur_prompt);
	if (retval != PAM_SUCCESS) {
	    return PAM_IGNORE;
	}
	if (cur_prompt == NULL) {
	    cur_prompt = "";
	}

	/* first read in the issue file */

	if (parse_esc) {
	    prompt_tmp = do_prompt(fd);
	    if (prompt_tmp == NULL) {
		return PAM_IGNORE;
	    }
	} else {
	    int count = 0;

	    prompt_tmp = malloc(st.st_size + 1);
	    if (prompt_tmp == NULL) {
		return PAM_IGNORE;
	    }
	    memset (prompt_tmp, '\0', st.st_size + 1);
	    count = fread(prompt_tmp, sizeof(char *), st.st_size, fd);
	    if (count != st.st_size) {
		free(prompt_tmp);
		return PAM_IGNORE;
	    }
	    prompt_tmp[st.st_size] = '\0';
	}

	fclose(fd);

	tot_size = strlen(prompt_tmp) + strlen(cur_prompt) + 1;

       /*
	* alloc some extra space for the original prompt
	* and postpend it to the buffer
	*/
	{
	    char *prompt_tmp_tmp = prompt_tmp;

	    prompt_tmp = realloc(prompt_tmp, tot_size);
	    if (prompt_tmp == NULL) {
		prompt_tmp = prompt_tmp_tmp;
		retval = PAM_IGNORE;
		goto cleanup;
	    }
	}

	strcpy(prompt_tmp+strlen(prompt_tmp), cur_prompt);

	prompt_tmp[tot_size] = '\0';

	retval = pam_set_item(pamh, PAM_USER_PROMPT,
			      (const char *) prompt_tmp);

    cleanup:
	free(issue_file);
	free(prompt_tmp);

    } else {
	D(("could not open issue_file: %s", issue_file));
	return PAM_IGNORE;
    }

    return retval;
}

PAM_EXTERN
int pam_sm_setcred(pam_handle_t *pamh, int flags, int argc,
                   const char **argv)
{
     return PAM_IGNORE;
}

char *do_prompt(FILE *fd)
{
    int c, size = 1024;
    char *issue = (char *)malloc(size);
    char buf[1024];
    struct utsname uts;

    if (issue == NULL || fd == NULL)
	return NULL;

    issue[0] = '\0'; /* zero this, for strcat to work on first buf */
    (void) uname(&uts);

    while ((c = getc(fd)) != EOF) {
	if (c == '\\') {
	    c = getc(fd);
	    switch (c) {
	      case 's':
		snprintf (buf, 1024, "%s", uts.sysname);
		break;
	      case 'n':
		snprintf (buf, 1024, "%s", uts.nodename);
		break;
	      case 'r':
		snprintf (buf, 1024, "%s", uts.release);
		break;
	      case 'v':
		snprintf (buf, 1024, "%s", uts.version);
		break;
	      case 'm':
		snprintf (buf, 1024, "%s", uts.machine);
		break;
	      case 'o':
		{
		    char domainname[256];

		    getdomainname(domainname, sizeof(domainname));
		    domainname[sizeof(domainname)-1] = '\0';
		    snprintf (buf, 1024, "%s", domainname);
		}
		break;

	      case 'd':
	      case 't':
		{
		    const char *weekday[] = {
			"Sun", "Mon", "Tue", "Wed", "Thu",
			"Fri", "Sat" };
		    const char *month[] = {
			"Jan", "Feb", "Mar", "Apr", "May",
			"Jun", "Jul", "Aug", "Sep", "Oct",
			"Nov", "Dec" };
		    time_t now;
		    struct tm *tm;

		    (void) time (&now);
		    tm = localtime(&now);

		    if (c == 'd')
			snprintf (buf, 1024, "%s %s %d  %d",
				weekday[tm->tm_wday], month[tm->tm_mon],
				tm->tm_mday, 
				tm->tm_year + 1900);
		    else
			snprintf (buf, 1024, "%02d:%02d:%02d",
				tm->tm_hour, tm->tm_min, tm->tm_sec);
		}
		break;
	      case 'l':
		{
		    char *ttyn = ttyname(1);
		    if (ttyn) {
		      if (!strncmp(ttyn, "/dev/", 5))
			ttyn += 5;
		      snprintf (buf, 1024, "%s", ttyn);
		    } else {
		      buf[0] = '\0';
		    }
		}
		break;
	      case 'u':
	      case 'U':
		{
		    int users = 0;
		    struct utmp *ut;
		    setutent();
		    while ((ut = getutent()))
			if (ut->ut_type == USER_PROCESS)
			users++;
		    endutent();
		    printf ("%d ", users);
		    if (c == 'U')
			snprintf (buf, 1024, "%s", (users == 1) ?
				" user" : " users");
		    break;
		}
	      default:
		buf[0] = c; buf[1] = '\0';
	    }
	    if ((strlen(issue) + strlen(buf)) < size + 1) {
		char *old_issue = issue;

		size += strlen(buf) + 1;
		issue = (char *) realloc (issue, size);
		if (issue == NULL) {
		    free(old_issue);
		    return NULL;
		}
	    }
	    strcat(issue, buf);
	} else {
	    buf[0] = c; buf[1] = '\0';
	    if ((strlen(issue) + strlen(buf)) < size + 1) {
		char *old_issue = issue;

		size += strlen(buf) + 1;
		issue = (char *) realloc (issue, size);
		if (issue == NULL) {
		    free(old_issue);
		    return NULL;
		}
	    }
	    strcat(issue, buf);
	}
    }

    return issue;
}

#ifdef PAM_STATIC

/* static module data */

struct pam_module _pam_issue_modstruct = {
     "pam_issue",
     pam_sm_authenticate,
     pam_sm_setcred,
     NULL,
     NULL,
     NULL,
     NULL,
};

#endif

/* end of module definition */
